documentation/source/coroutines.rst

```diff
@@ -151,3 +151,35 @@ they'd naturally end.
         if not isclose(edge_time_ns, start_time_ns + 4.0):
             raise TestFailure("Expected a period of 4 ns")
 
+
+Async functions
+---------------
+
+Python 3.5 introduces ``async`` functions, which provide an alternative syntax.
+For example:
+
+.. code-block:: python
+
+    @cocotb.coroutine
+    async def wait_10ns():
+        cocotb.log.info("About to wait for 10ns")
+        await Timer(10000)
+        cocotb.log.info("Simulation time has advanced by 10 ns")
+
+To wait on a trigger or a nested coroutine, these use ``await`` instead of ``yield``.
+
+This frees up ``yield`` for its indended use in iterators, which python 3.6
+enables via ``async for``:
+
+.. code-block:: python
+
+    async def ten_samples_of(clk, signal):
+        for i in range(10):
+            await RisingEdge(clk)
+            yield signal.value
+
+
+    @cocotb.coroutine
+    async def check_some_samples(dut):
+        async for sample in samples_of(dut.clk, dut.signal):
+            assert sample % 2 == 0
```

Comment:
You could make the docs even clearer by adding a sentence or two on what this code does, and/or rename `check_some_samples` to `check_samples_are_even` (or something like that)

Reply:
Can you think of a less contrived example I could give here instead? Iterating over axi-lite memory is one that comes to mind, but that's a little more involved.

Reply:
(names updated anyway)

Reply:
I think the example is simple enough for people to understand what's going on, let's leave it as-is.